lib/core/src/Cardano/Wallet/Primitive/Migration/Planning.hs

```diff
@@ -0,0 +1,307 @@
+data MigrationPlan input size = MigrationPlan
+    { selections :: ![Selection input size]
+      -- ^ A list of generated selections: each selection is the basis for a
+      -- single transaction.
+    , unselected :: !(CategorizedUTxO input)
+      -- ^ The portion of the UTxO that was not selected.
+    , totalFee :: !Coin
+      -- ^ The total fee payable: equal to the sum of the fees of the
+      -- individual selections.
+    }
+    deriving (Eq, Generic, Show)
+
+-- | Creates a migration plan for the given categorized UTxO set and reward
+--   withdrawal amount.
+--
+-- See 'MigrationPlan'.
+--
+createPlan
+    :: TxSize size
+    => TxConstraints size
+    -> CategorizedUTxO input
+    -> RewardWithdrawal
+    -> MigrationPlan input size
+createPlan constraints =
+    run []
+  where
+    run !selections !utxo !reward =
+        case createSelection constraints utxo reward of
+            Just (utxo', selection) ->
+                run (selection : selections) utxo' (RewardWithdrawal $ Coin 0)
+            Nothing -> MigrationPlan
+                { selections
+                , unselected = utxo
+                , totalFee = F.foldMap (view #fee) selections
+                }
+
+-- | Creates an individual selection for inclusion in a migration plan.
+--
+-- A selection is the basis for an individual transaction.
+--
+-- Returns 'Nothing' if it was not possible to create a selection with the UTxO
+-- entries that remain.
+--
+createSelection
+    :: TxSize size
+    => TxConstraints size
+    -> CategorizedUTxO input
+    -> RewardWithdrawal
+    -> Maybe (CategorizedUTxO input, Selection input size)
+createSelection constraints utxo rewardWithdrawal =
+    initializeSelection constraints utxo rewardWithdrawal
+    <&> extendSelection constraints
+
+-- | Initializes a selection with a single entry.
+--
+-- Returns 'Nothing' if it was not possible to initialize a selection with the
+-- UTxO entries that remain.
+--
+initializeSelection
+    :: forall input size. TxSize size
+    => TxConstraints size
+    -> CategorizedUTxO input
+    -> RewardWithdrawal
+    -> Maybe (CategorizedUTxO input, Selection input size)
+initializeSelection constraints utxoAtStart reward =
+    initializeWith =<< utxoAtStart `select` Supporter
+  where
+    initializeWith (entry, utxo) =
+        case Selection.create constraints reward [entry] of
+            Right selection -> Just (utxo, selection)
+            Left _ -> Nothing
+
+-- | Extends a selection repeatedly, until the selection is full.
+--
+-- This function terminates when the selection cannot be extended further
+-- (because doing so would cause it to exceed the size limit of a transaction),
+-- or when there are no more UTxO entries available for selection.
+--
+-- Priority is given to selecting "freerider" entries: entries that cannot pay
+-- for themselves. A "supporter" entry is only added to the selection if there
+-- is not enough ada to pay for a "freerider" entry.
+--
+extendSelection
```

Comment:
Thanks for the suggestion to rename this. I agree it could be a little clearer.

I've renamed this to `extendSelectionUntilFull`, as it does indeed repeatedly extend a selection until it's full.

Fixed in 97f5c261b6d23b65e14dcb67c72101a62db8b58d.